libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp

```diff
@@ -29,352 +33,109 @@
-            if (mpi_status == MPI_SUCCESS)
-            {
-                pika::execution::experimental::set_value(
-                    PIKA_FORWARD(Receiver, receiver), PIKA_FORWARD(Ts, ts)...);
-            }
-            else
-            {
-                pika::execution::experimental::set_error(PIKA_FORWARD(Receiver, receiver),
-                    std::make_exception_ptr(mpi_exception(mpi_status)));
-            }
-        }
-
-        template <typename OperationState>
-        void set_value_request_callback_void(MPI_Request request, OperationState& op_state)
-        {
-            detail::add_request_callback(
-                [&op_state](int status) mutable {
-                    op_state.ts = {};
-                    set_value_request_callback_helper(status, PIKA_MOVE(op_state.receiver));
-                },
-                request, op_state.stream);
-        }
-
-        template <typename Result, typename OperationState>
-        void set_value_request_callback_non_void(MPI_Request request, OperationState& op_state)
-        {
-            detail::add_request_callback(
-                [&op_state](int status) mutable {
-                    op_state.ts = {};
-                    PIKA_ASSERT(std::holds_alternative<Result>(op_state.result));
-                    set_value_request_callback_helper(status, PIKA_MOVE(op_state.receiver),
-                        PIKA_MOVE(std::get<Result>(op_state.result)));
-                },
-                request, op_state.stream);
-        }
-
-        template <typename F, typename... Ts>
-        inline constexpr bool is_mpi_request_invocable_v =
-            std::is_invocable_v<F, std::add_lvalue_reference_t<std::decay_t<Ts>>..., MPI_Request*>;
-
-        template <typename F, typename... Ts>
-        using mpi_request_invoke_result_t = std::decay_t<std::invoke_result_t<F,
-            std::add_lvalue_reference_t<std::decay_t<Ts>>..., MPI_Request*>>;
-
-        template <typename Sender, typename F>
-        struct transform_mpi_sender_impl
-        {
-            struct transform_mpi_sender_type;
-        };
 
-        template <typename Sender, typename F>
-        using transform_mpi_sender =
-            typename transform_mpi_sender_impl<Sender, F>::transform_mpi_sender_type;
+    namespace pud = pika::util::detail;
+    namespace exp = execution::experimental;
 
-        template <typename Sender, typename F>
-        struct transform_mpi_sender_impl<Sender, F>::transform_mpi_sender_type
+    inline constexpr struct transform_mpi_t final
+      : pika::functional::detail::tag_fallback<transform_mpi_t>
+    {
+    private:
+        template <typename Sender, typename F,
+            PIKA_CONCEPT_REQUIRES_(exp::is_sender_v<std::decay_t<Sender>>)>
+        friend constexpr PIKA_FORCEINLINE exp::unique_any_sender<>
+        tag_fallback_invoke(transform_mpi_t, Sender&& sender, F&& f,
+            /*progress_mode p, */ stream_type s = stream_type::automatic)
```

Comment:
What is the `progress_mode` meant to do (eventually) and does it need to stay around?

Reply:
removed for now